File: web/src/lib/appeal/coaching.ts
/**
 * Appeal coaching helpers — keep in sync with src/appeal/coaching.py (canonical).
 */
import { JUDGE_ORDER } from "../sse/types.ts";
import type { JudgeId, Verdict, VerdictLabel } from "../sse/types.ts";

const APPEAL_PRIORITY: Record<VerdictLabel, number> = {
  FAIL: 0,
  CONDITIONAL: 1,
  PASS: 2,
};

const DERIVED_HINT_PREFIX = "Provide concrete evidence that addresses:";

// ponytail: Jaccard on word tokens catches near-paraphrase; upgrade path is embedding similarity.
const LENS_SIMILARITY_THRESHOLD = 0.85;

const GENERIC_EVIDENCE_PHRASES = [
  "do more research",
  "conduct more research",
  "provide more evidence",
  "show more evidence",
  "gather more data",
  "need more validation",
  "validate the market",
  "show traction",
  "prove product-market fit",
  "prove product market fit",
  "conduct market research",
  "more customer discovery",
  "build a stronger case",
] as const;

const FILLER_AFTER_GENERIC_PHRASE = new Set([
  "on",
  "the",
  "a",
  "an",
  "this",
  "that",
  "idea",
  "buyer",
  "market",
  "please",
  "first",
  "more",
  "your",
  "and",
  "or",
  "to",
  "for",
]);

function normalizeSentence(text: string): string {
  return text.toLowerCase().split(/\s+/).filter(Boolean).join(" ");
}

export function sentenceSimilarity(left: string, right: string): number {
  const leftTokens = new Set(normalizeSentence(left).split(" "));
  const rightTokens = new Set(normalizeSentence(right).split(" "));
  if (!leftTokens.size || !rightTokens.size) return 0;
  let overlap = 0;
  for (const token of leftTokens) {
    if (rightTokens.has(token)) overlap += 1;
  }
  const union = new Set([...leftTokens, ...rightTokens]).size;
  return union ? overlap / union : 0;
}

function isGenericClause(normalized: string): boolean {
  const phrases = [...GENERIC_EVIDENCE_PHRASES].sort(
    (left, right) => right.length - left.length,
  );
  for (const phrase of phrases) {
    if (normalized === phrase) return true;
    if (normalized.startsWith(phrase)) {
      const rest = normalized.slice(phrase.length).trim().replace(/[.]+$/, "");
      if (!rest || rest.split(/\s+/).every((word) => FILLER_AFTER_GENERIC_PHRASE.has(word))) {
        return true;
      }
    }
  }
  return false;
}

export function isGenericEvidence(text: string): boolean {
  const normalized = normalizeSentence(text);
  if (!normalized) return true;
  const parts = normalized
    .split(" and ")
    .map((part) => part.trim())
    .filter(Boolean);
  if (parts.length > 1) {
    return parts.every((part) => isGenericClause(part));
  }
  return isGenericClause(normalized);
}

export type AppealHintQuality = "precise" | "derived" | "generic" | "duplicate";

export function isDerivedCoachingHint(hint: string): boolean {
  return hint.trim().startsWith(DERIVED_HINT_PREFIX);
}

export function appealCoachingHint(verdict: Verdict): string {
  const evidence = verdict.evidence_to_change_verdict?.trim();
  if (evidence) return evidence;
  return `${DERIVED_HINT_PREFIX} ${verdict.key_concern.trim()}`;
}

export function isDegenerateEvidenceAsks(verdicts: Verdict[]): boolean {
  const asks = verdicts
    .map((verdict) => normalizeSentence(appealCoachingHint(verdict)))
    .filter(Boolean);
  if (asks.length < 2) return false;
  const first = asks[0];
  return asks.every((ask) => ask === first);
}

function hintQuality(
  verdict: Verdict,
  duplicateJudges: Set<JudgeId>,
): AppealHintQuality {
  if (duplicateJudges.has(verdict.judge)) return "duplicate";
  const evidence = verdict.evidence_to_change_verdict?.trim();
  if (!evidence) return "derived";
  if (isGenericEvidence(evidence)) return "generic";
  return "precise";
}

export function appealCoachingVerdicts(verdicts: Verdict[]): Verdict[] {
  return [...verdicts].sort((left, right) => {
    const priority =
      APPEAL_PRIORITY[left.verdict] - APPEAL_PRIORITY[right.verdict];
    if (priority !== 0) return priority;
    return (
      JUDGE_ORDER.indexOf(left.judge) - JUDGE_ORDER.indexOf(right.judge)
    );
  });
}

/** Verdicts with enough data to coach or appeal against. */
export function appealBaselineVerdicts(verdicts: Verdict[]): Verdict[] {
  return verdicts.filter((verdict) => verdict.score > 0 || Boolean(verdict.roast));
}

export interface AppealCoachingItem {
  judge: JudgeId;
  hint: string;
  verdict: VerdictLabel;
  score: number;
  quality: AppealHintQuality;
}

export interface AppealCoachingAssessment {
  degraded: boolean;
  reasons: string[];
  items: AppealCoachingItem[];
  degenerateAsks: boolean;
}

/** Judges whose normalized evidence ask collides with another panel member. */
export function findDuplicateEvidenceJudges(verdicts: Verdict[]): Set<JudgeId> {
  const duplicateJudges = new Set<JudgeId>();
  const seen = new Map<string, JudgeId>();

  for (const verdict of verdicts) {
    const normalized = normalizeSentence(appealCoachingHint(verdict));
    if (!normalized) continue;
    const prior = seen.get(normalized);
    if (prior !== undefined) {
      duplicateJudges.add(verdict.judge);
      duplicateJudges.add(prior);
    } else {
      seen.set(normalized, verdict.judge);
    }
  }

  const evidenceItems: { judge: JudgeId; text: string }[] = [];
  for (const verdict of verdicts) {
    const evidence = verdict.evidence_to_change_verdict?.trim();
    if (!evidence) continue;
    evidenceItems.push({
      judge: verdict.judge,
      text: normalizeSentence(evidence),
    });
  }
  for (let index = 0; index < evidenceItems.length; index += 1) {
    for (let other = index + 1; other < evidenceItems.length; other += 1) {
      const left = evidenceItems[index];
      const right = evidenceItems[other];
      if (
        left.text === right.text ||
        sentenceSimilarity(left.text, right.text) >= LENS_SIMILARITY_THRESHOLD
      ) {
        duplicateJudges.add(left.judge);
        duplicateJudges.add(right.judge);
      }
    }
  }

  return duplicateJudges;
}

export function assessAppealCoaching(verdicts: Verdict[]): AppealCoachingAssessment {
  const ordered = appealCoachingVerdicts(verdicts);
  const duplicateJudges = findDuplicateEvidenceJudges(ordered);

  const items: AppealCoachingItem[] = ordered.map((verdict) => ({
    judge: verdict.judge,
    hint: appealCoachingHint(verdict),
    verdict: verdict.verdict,
    score: verdict.score,
    quality: hintQuality(verdict, duplicateJudges),
  }));

  const degenerateAsks = ordered.length > 0 && isDegenerateEvidenceAsks(ordered);
  const reasons: string[] = [];
  if (degenerateAsks) {
    reasons.push("Judges returned near-identical evidence asks.");
  } else if (duplicateJudges.size > 0) {
    reasons.push("Some judges asked for the same proof.");
  } else {
    const genericCount = items.filter((item) => item.quality === "generic").length;
    if (genericCount > 0) {
      reasons.push(
        `${genericCount} evidence ask${genericCount === 1 ? "" : "s"} look generic — treat them as directional, not precise targets.`,
      );
    }
  }

  return {
    degraded: reasons.length > 0,
    reasons,
    items,
    degenerateAsks,
  };
}

export function normalizeTargetJudges(judges: JudgeId[] | undefined): JudgeId[] {
  if (!judges?.length) return [];
  const allowed = new Set(JUDGE_ORDER);
  return JUDGE_ORDER.filter((judge) => allowed.has(judge) && judges.includes(judge));
}

export function appealEvidenceOutcome(
  original: Verdict,
  revised: Verdict,
): string {
  const delta = revised.score - original.score;
  if (delta > 0) return "Evidence met";
  if (delta < 0) return "Not met";
  if (original.verdict === "PASS") return "Already passing";
  return "Not met";
}

export interface AppealJudgeOutcome {
  judge: JudgeId;
  evidenceAsk: string;
  outcome: string;
  targeted: boolean;
  scoreDelta: number;
}

export function appealJudgeOutcomes(
  baseline: Verdict[],
  revised: Verdict[],
  targetJudges: JudgeId[] = [],
): AppealJudgeOutcome[] {
  const originals = new Map(baseline.map((verdict) => [verdict.judge, verdict]));
  const targets = new Set(normalizeTargetJudges(targetJudges));
  return revised.flatMap((revisedVerdict) => {
    const original = originals.get(revisedVerdict.judge);
    if (!original) return [];
    const scoreDelta = revisedVerdict.score - original.score;
    return [
      {
        judge: revisedVerdict.judge,
        evidenceAsk: appealCoachingHint(original),
        outcome: appealEvidenceOutcome(original, revisedVerdict),
        targeted: targets.has(revisedVerdict.judge),
        scoreDelta,
      },
    ];
  });
}

export function appealScoreMovement(
  baseline: Verdict[],
  revised: Verdict[],
): { positiveMoves: number; netDelta: number } {
  const originals = new Map(baseline.map((verdict) => [verdict.judge, verdict]));
  let positiveMoves = 0;
  let netDelta = 0;
  for (const revisedVerdict of revised) {
    const original = originals.get(revisedVerdict.judge);
    if (!original) continue;
    const delta = revisedVerdict.score - original.score;
    netDelta += delta;
    if (delta > 0) positiveMoves += 1;
  }
  return { positiveMoves, netDelta };
}

File: src/appeal/coaching.py
"""Appeal coaching hints derived from panel verdicts.

Keep in sync with web/src/lib/appeal/coaching.ts (client-side fallback).

"""

from dataclasses import dataclass
from typing import Literal

from config import JUDGE_ORDER
from judges.schemas import RoastPanel, Verdict, VerdictLabel
from verification.invariants import is_generic_evidence, normalize_sentence
from verification.lens import DERIVED_HINT_PREFIX, coaching_hint, find_duplicate_evidence_judges

_APPEAL_PRIORITY = {
    VerdictLabel.FAIL: 0,
    VerdictLabel.CONDITIONAL: 1,
    VerdictLabel.PASS: 2,
}

AppealHintQuality = Literal["precise", "derived", "generic", "duplicate"]


def is_derived_coaching_hint(hint: str) -> bool:

    return hint.strip().startswith(DERIVED_HINT_PREFIX)


def is_degenerate_evidence_asks(verdicts: list[Verdict]) -> bool:

    asks = [
        normalize_sentence(appeal_coaching_hint(verdict))
        for verdict in verdicts
        if appeal_coaching_hint(verdict).strip()
    ]

    if len(asks) < 2:
        return False

    first = asks[0]

    return all(ask == first for ask in asks[1:])


def _hint_quality(
    verdict: Verdict,
    *,
    duplicate_judges: set[str],
) -> AppealHintQuality:

    judge = verdict.judge.value

    if judge in duplicate_judges:
        return "duplicate"

    evidence = (verdict.evidence_to_change_verdict or "").strip()

    if not evidence:
        return "derived"

    if is_generic_evidence(evidence):
        return "generic"

    return "precise"


def appeal_coaching_hint(verdict: Verdict) -> str:
    return coaching_hint(verdict)


def appeal_coaching_verdicts(panel: RoastPanel) -> list[Verdict]:

    return sorted(
        panel.verdicts,
        key=lambda verdict: (_APPEAL_PRIORITY.get(verdict.verdict, 99), verdict.judge.value),
    )


@dataclass(frozen=True)
class AppealCoachingItem:
    judge: str

    hint: str

    verdict: str

    score: int

    quality: AppealHintQuality


def assess_appeal_coaching(panel: RoastPanel) -> dict:
    """Flag weak or duplicated evidence asks so coaching stays trustworthy."""

    ordered = appeal_coaching_verdicts(panel)
    duplicate_judges = find_duplicate_evidence_judges(ordered)

    items = [
        AppealCoachingItem(
            judge=verdict.judge.value,
            hint=appeal_coaching_hint(verdict),
            verdict=verdict.verdict.value,
            score=verdict.score,
            quality=_hint_quality(verdict, duplicate_judges=duplicate_judges),
        )
        for verdict in ordered
    ]

    degenerate_asks = is_degenerate_evidence_asks(ordered) if ordered else False

    reasons: list[str] = []

    if degenerate_asks:
        reasons.append("Judges returned near-identical evidence asks.")

    elif duplicate_judges:
        reasons.append("Some judges asked for the same proof.")

    elif any(item.quality == "generic" for item in items):
        generic_count = sum(1 for item in items if item.quality == "generic")

        reasons.append(
            f"{generic_count} evidence ask{'s' if generic_count != 1 else ''} "
            "look generic — treat them as directional, not precise targets."
        )

    return {
        "degraded": bool(reasons),
        "reasons": reasons,
        "items": items,
        "degenerate_asks": degenerate_asks,
    }


def appeal_score_movement(baseline: RoastPanel, revised: RoastPanel) -> dict[str, int]:
    """Score movement summary for appeal evals."""

    originals = {verdict.judge.value: verdict for verdict in baseline.verdicts}

    positive_moves = 0

    net_delta = 0

    for revised_verdict in revised.verdicts:
        original = originals.get(revised_verdict.judge.value)

        if original is None:
            continue

        delta = revised_verdict.score - original.score

        net_delta += delta

        if delta > 0:
            positive_moves += 1

    return {
        "positive_moves": positive_moves,
        "net_delta": net_delta,
    }


def normalize_target_judges(judges: list[str] | None) -> tuple[str, ...]:
    """Keep only known judge ids, in panel order."""

    if not judges:
        return ()

    allowed = set(JUDGE_ORDER)

    return tuple(judge for judge in JUDGE_ORDER if judge in allowed and judge in judges)


def appeal_evidence_outcome(original: Verdict, revised: Verdict) -> str:
    """Whether the original evidence ask was satisfied by the appeal."""

    delta = revised.score - original.score

    if delta > 0:
        return "Evidence met"

    if delta < 0:
        return "Not met"

    if original.verdict == VerdictLabel.PASS:
        return "Already passing"

    return "Not met"


@dataclass(frozen=True)
class AppealJudgeOutcome:
    judge: str

    evidence_ask: str

    outcome: str

    targeted: bool

    score_delta: int


def appeal_judge_outcomes(
    baseline: RoastPanel,
    revised: RoastPanel,
    target_judges: tuple[str, ...] = (),
) -> list[AppealJudgeOutcome]:
    """Per-judge evidence ask, founder targeting, and post-appeal outcome."""

    originals = {verdict.judge.value: verdict for verdict in baseline.verdicts}

    outcomes: list[AppealJudgeOutcome] = []

    for revised_verdict in revised.verdicts:
        judge = revised_verdict.judge.value

        original = originals.get(judge)

        if original is None:
            continue

        delta = revised_verdict.score - original.score

        outcomes.append(
            AppealJudgeOutcome(
                judge=judge,
                evidence_ask=appeal_coaching_hint(original),
                outcome=appeal_evidence_outcome(original, revised_verdict),
                targeted=judge in target_judges,
                score_delta=delta,
            )
        )

    return outcomes
